File: agent/src/agent_runner/runner.py
"""Claude Code SDK を使って Research Agent を実行する"""

import logging

from claude_code_sdk import ClaudeCodeOptions, ResultMessage, AssistantMessage, TextBlock, query
import claude_code_sdk._internal.message_parser as _mp
import claude_code_sdk._internal.client as _client

from agent_runner.config import settings
from agent_runner.markets import MARKETS

logger = logging.getLogger(__name__)

# SDK が未知のメッセージタイプ (rate_limit_event 等) で crash するのを防ぐ
_original_parse = _mp.parse_message


def _patched_parse(data):
    try:
        return _original_parse(data)
    except _mp.MessageParseError as e:
        if "Unknown message type" in str(e):
            logger.debug("Skipping unknown message type: %s", data.get("type"))
            return None
        raise


_mp.parse_message = _patched_parse
_client.parse_message = _patched_parse


def _build_mcp_servers() -> dict:
    """MCP サーバー設定を構築する"""
    return {
        "claude-trade": {
            "type": "stdio",
            "command": settings.mcp_server_command,
            "args": [
                "run",
                "--directory", settings.mcp_server_dir,
                "python", "-m", "claude_trade.server",
            ],
        }
    }


def _load_prompt(mode: str, market: str = "us") -> str:
    """プロンプトファイルを読み込み、マーケットコンテキストを付加する"""
    prompt_map = {
        "premarket": "premarket-analysis.md",
        "intraday": "research-loop.md",
        "eod": "eod-review.md",
    }
    filename = prompt_map.get(mode)
    if not filename:
        raise ValueError(f"Unknown mode: {mode}. Use: {list(prompt_map.keys())}")

    prompt_path = settings.prompts_dir / filename
    if not prompt_path.exists():
        raise FileNotFoundError(f"Prompt file not found: {prompt_path}")

    base_prompt = prompt_path.read_text(encoding="utf-8")

    # マーケット固有の戦略を読み込み、プロンプトに付加
    mkt = MARKETS[market]
    strategy_path = settings.strategies_dir / mkt.strategy_file
    strategy_content = strategy_path.read_text(encoding="utf-8")

    market_context = f"""

## 対象マーケット: {mkt.name}

- マーケットID: `{mkt.market_id}`
- タイムゾーン: {mkt.timezone}
- 取引所: {mkt.ib_exchange}
- 通貨: {mkt.ib_currency}
- シグナルの `source_strategy` には `"{mkt.market_id}"` を指定してください

## 戦略定義

```yaml
{strategy_content}
```

## MCP ツールの注意事項

- `get_quote`, `get_historical_data` には `exchange="{mkt.ib_exchange}"`, `currency="{mkt.ib_currency}"` を渡してください
- `write_signal` の `strategy` には `"{mkt.market_id}"` を指定してください

## フィードバックループ

- 分析開始時に `get_relevant_lessons(source_strategy="{mkt.market_id}")` で過去の学びを確認してください
- 過去の学びに基づいて分析の重点を調整してください
- `get_signal_accuracy(source_strategy="{mkt.market_id}")` でシグナル精度の傾向を把握してください
- EOD Review 時は `evaluate_signal` で各シグナルの事後評価を、`record_lesson` で学びの記録を行ってください
"""
    return base_prompt + market_context


async def run_research(mode: str, market: str = "us") -> dict:
    """Research Agent を実行する

    Args:
        mode: "premarket", "intraday", or "eod"
        market: "us", "jp", or "eu"

    Returns:
        実行結果のサマリー
    """
    prompt = _load_prompt(mode, market)
    logger.info("Starting %s research for %s market", mode, market.upper())

    options = ClaudeCodeOptions(
        allowed_tools=["mcp__claude-trade__*", "WebSearch", "WebFetch"],
        mcp_servers=_build_mcp_servers(),
        model=settings.model,
        max_turns=settings.max_turns,
        permission_mode="bypassPermissions",
        cwd=str(settings.project_root),
    )

    result_text = ""
    session_id = ""
    total_cost = None

    try:
        async for message in query(prompt=prompt, options=options):
            if message is None:
                continue
            if isinstance(message, AssistantMessage):
                for block in message.content:
                    if isinstance(block, TextBlock):
                        logger.info("[%s/%s] %s", market.upper(), mode, block.text[:200])
            elif isinstance(message, ResultMessage):
                result_text = message.result or ""
                session_id = message.session_id
                total_cost = message.total_cost_usd
                logger.info(
                    "[%s/%s] Completed: turns=%d, cost=$%s, duration=%dms",
                    market.upper(), mode,
                    message.num_turns,
                    message.total_cost_usd,
                    message.duration_ms,
                )
    except Exception:
        logger.exception("[%s/%s] SDK error during research", market.upper(), mode)

    return {
        "mode": mode,
        "market": market,
        "session_id": session_id,
        "result": result_text,
        "total_cost_usd": total_cost,
    }

File: packages/server/src/agent/runner.ts
/** Claude Code SDK を使って Research Agent を実行する */

import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import {
  type McpStdioServerConfig,
  type SDKAssistantMessage,
  type SDKMessage,
  type SDKResultMessage,
  query,
} from "@anthropic-ai/claude-code";
import { PROJECT_ROOT, loadEnv } from "../config.js";
import { createLogger } from "../lib/logger.js";
import { MARKETS } from "./markets.js";

const env = loadEnv();
const PROMPTS_DIR = resolve(PROJECT_ROOT, "prompts");
const STRATEGIES_DIR = resolve(PROJECT_ROOT, "strategies");

function buildMcpServers(): Record<string, McpStdioServerConfig> {
  return {
    "claude-trade": {
      type: "stdio",
      command: "npx",
      args: ["tsx", resolve(PROJECT_ROOT, "packages/server/src/mcp-server/server.ts")],
    },
  };
}

function loadPrompt(mode: string, market = "us"): string {
  const promptMap: Record<string, string> = {
    premarket: "premarket-analysis.md",
    intraday: "research-loop.md",
    eod: "eod-review.md",
  };

  const filename = promptMap[mode];
  if (!filename) {
    throw new Error(`Unknown mode: ${mode}. Use: ${Object.keys(promptMap).join(", ")}`);
  }

  const promptPath = resolve(PROMPTS_DIR, filename);
  const basePrompt = readFileSync(promptPath, "utf-8");

  const mkt = MARKETS[market];
  if (!mkt) throw new Error(`Unknown market: ${market}`);

  const strategyContent = readFileSync(resolve(STRATEGIES_DIR, mkt.strategyFile), "utf-8");

  const marketContext = `

## 対象マーケット: ${mkt.name}

- マーケットID: \`${mkt.marketId}\`
- タイムゾーン: ${mkt.timezone}
- 取引所: ${mkt.ibExchange}
- 通貨: ${mkt.ibCurrency}
- シグナルの \`source_strategy\` には \`"${mkt.marketId}"\` を指定してください

## 戦略定義

\`\`\`yaml
${strategyContent}
\`\`\`

## MCP ツールの注意事項

- \`get_quote\`, \`get_historical_data\` には \`exchange="${mkt.ibExchange}"\`, \`currency="${mkt.ibCurrency}"\` を渡してください
- \`write_signal\` の \`strategy\` には \`"${mkt.marketId}"\` を指定してください

## フィードバックループ

- 分析開始時に \`get_relevant_lessons(source_strategy="${mkt.marketId}")\` で過去の学びを確認してください
- 過去の学びに基づいて分析の重点を調整してください
- \`get_signal_accuracy(source_strategy="${mkt.marketId}")\` でシグナル精度の傾向を把握してください
- EOD Review 時は \`evaluate_signal\` で各シグナルの事後評価を、\`record_lesson\` で学びの記録を行ってください
`;

  return basePrompt + marketContext;
}

export async function runResearch(
  mode: string,
  market = "us",
): Promise<{
  mode: string;
  market: string;
  sessionId: string;
  result: string;
  totalCostUsd: number | null;
}> {
  const log = createLogger("research", market.toUpperCase(), mode);
  const prompt = loadPrompt(mode, market);
  log.info("Starting research");

  let resultText = "";
  let sessionId = "";
  let totalCost: number | null = null;

  try {
    const conversation = query({
      prompt,
      options: {
        allowedTools: ["mcp__claude-trade__*", "WebSearch", "WebFetch"],
        mcpServers: buildMcpServers(),
        model: env.MODEL,
        maxTurns: env.MAX_TURNS,
        permissionMode: "bypassPermissions",
        cwd: PROJECT_ROOT,
      },
    });

    for await (const message of conversation) {
      if (message == null) continue;

      const msg = message as SDKMessage;
      if (msg.type === "assistant") {
        const assistantMsg = msg as SDKAssistantMessage;
        if (assistantMsg.message?.content) {
          for (const block of assistantMsg.message.content) {
            if (block.type === "text" && "text" in block) {
              log.info((block as { text: string }).text.slice(0, 200));
            }
          }
        }
      } else if (msg.type === "result") {
        const resultMsg = msg as SDKResultMessage;
        if ("result" in resultMsg) {
          resultText = resultMsg.result ?? "";
        }
        sessionId = resultMsg.session_id;
        totalCost = resultMsg.total_cost_usd;
        log.info(
          {
            turns: resultMsg.num_turns,
            cost: resultMsg.total_cost_usd,
            durationMs: resultMsg.duration_ms,
          },
          "Research completed",
        );
      }
    }
  } catch (e) {
    log.error({ err: e }, "SDK error during research");
  }

  return {
    mode,
    market,
    sessionId,
    result: resultText,
    totalCostUsd: totalCost,
  };
}
